src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+{
+    /// <summary>
+    /// Provider that computes display order of tree items based on input ordering of
+    /// evaluated includes from the project file.
+    /// </summary>
+    internal class TreeItemOrderPropertyProvider : IProjectTreePropertiesProvider
+    {
+        private const string FullPathProperty = "FullPath";
+        private readonly char[] _separators = new char[]
+        {
+            Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar
+        };
+        private Dictionary<string, int> _displayOrderMap = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+        private Dictionary<string, int> _rootedOrderMap = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+        private UnconfiguredProject _project;
+        private ImmutableHashSet<string> _allowedItemTypes;
+
+        public TreeItemOrderPropertyProvider(IReadOnlyCollection<ProjectItemIdentity> orderedItems, UnconfiguredProject project)
+        {
+            _project = project;
+            OrderedItems = orderedItems;
+
+            _allowedItemTypes = OrderedItems.Select(p => p.ItemType).ToImmutableHashSet();
+
+            ComputeIndices();
+        }
+
+        public IReadOnlyCollection<ProjectItemIdentity> OrderedItems { get; }
+
+        /// <summary>
+        /// Preorder folders and items that are provided as ordered evaluated includes
+        /// </summary>
+        private void ComputeIndices()
+        {
+            var duplicateFiles = OrderedItems
+                .Select(p => Path.GetFileName(p.EvaluatedInclude))
+                .GroupBy(file => file, StringComparer.OrdinalIgnoreCase)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key).ToImmutableHashSet();
+
+            var index = 1;
+            
+            foreach (var item in OrderedItems)
```

Comment:
Can you add a comment here describing what this results in? I've read this code a few times and having real trouble picturing what is occurring.

Reply:
A folder structure that I'm thinking will be tripped by this:

```
Folder.fs\Folder\Bar.fs
Folder\Folder.fs
```